kvclientbabysitter.go

```diff
@@ -0,0 +1,559 @@
+	// if there is no current client, we obviously can't use it
+	if p.currentClient == nil {
+		p.activeClient = nil
+		return
+	}
+
+	// if we previously had no bucket selected, but now do, we cannot use
+	// the client to prevent a race condition where the bucket is not
+	// actually selected yet.  Note that the converse is not true, if we
+	// had a bucket selected, but now do not, that is safe to use still.
+	if p.currentConfig.SelectedBucket == "" && p.desiredConfig.SelectedBucket != "" {
+		p.activeClient = nil
+		return
+	}
+
+	p.activeClient = p.currentClient
+}
+
+func (p *kvClientBabysitter) rebuildFastLookupLocked() {
+	p.client.Store(&kvClientBabysitterState{
+		Err:    p.connectErr,
+		Client: p.activeClient,
+	})
+}
+
+func (p *kvClientBabysitter) maybeBeginClientBuildLocked() {
+	if p.isClosed {
+		return
+	}
+	if p.isBuilding {
+		return
+	}
+
+	if p.activeClient != nil &&
+		p.currentConfig.Equals(*p.currentConfig) {
+		// already have a client with the desired config
+		return
+	}
+
+	buildCtx, buildCancelFn := context.WithCancel(context.Background())
+	buildDoneCh := make(chan struct{}, 1)
+
+	p.isBuilding = true
+	p.buildCancelFn = buildCancelFn
+	p.buildDoneCh = buildDoneCh
+
+	go func() {
+		p.clientBuildThread(buildCtx)
+
+		p.lock.Lock()
+		p.isBuilding = false
+		p.buildCancelFn = nil
+		p.buildDoneCh = nil
+		p.lock.Unlock()
+
+		buildCancelFn()
+		close(buildDoneCh)
+	}()
+}
+
+func (p *kvClientBabysitter) clientBuildThread(
+	ctx context.Context,
+) {
+	p.logger.Info("client build thread started")
+
+	lastErrTime := time.Time{}
+
+	// we are the only writer to these values, so it is safe to read them
+	// without a lock, and we do not need to refresh them while we are working.
+	currentConfig := p.currentConfig
+	currentClient := p.currentClient
+
+	p.lock.Lock()
+	desiredConfig := p.desiredConfig
+	p.lock.Unlock()
+
+ClientBuildLoop:
+	for {
+		if desiredConfig == nil {
+			p.logger.DPanic("desired config is nil in client build thread")
+		}
+		if currentClient != nil && currentConfig == nil {
+			p.logger.DPanic("current client is non-nil but current config is nil in client build thread")
+		}
+
+		if ctx.Err() != nil {
+			p.logger.Debug("client build thread exiting due to context done", zap.Error(ctx.Err()))
+			return
+		}
+
+		if currentConfig != nil && desiredConfig.Equals(*currentConfig) {
+			// we have reached the desired config
+			break
+		}
+
+		if currentClient != nil {
+			bucketChangeConfig := *currentConfig
+			bucketChangeConfig.SelectedBucket = desiredConfig.SelectedBucket
+			if desiredConfig.Equals(bucketChangeConfig) {
+				// if changing the bucket is enough to match the desired config, do that
+
+				err := currentClient.SelectBucket(ctx, desiredConfig.SelectedBucket)
+				if err == nil {
+					currentConfig = desiredConfig
+
+					p.lock.Lock()
+					p.currentConfig = currentConfig
+					p.updateActiveClientLocked()
+					p.rebuildFastLookupLocked()
+					p.lock.Unlock()
+					continue
```

Comment:
Can we just break here since we know desiredConfig == bucketChangeConfig == currentConfig?

Reply:
I think that would leave a race if another config change was received.